### ISSUE TITLE
Loss of ability to use advanced test setup since GHC 8 setup

### ISSUE THREAD

**Issue by Voileexperiments** [Codewars/codewars-runner-cli#655]
- [x] ~~Hspec with GHC 8 has 2 level deeper nesting~~ Removed "Tests". Now it just includes the name of the module it's testing.
- [x] ~~Can't do multiple top level blocks~~ multiple `describe`s are allowed
- [x] ~~Can't do something before running tests and use the result~~ `beforeAll`
- [ ] Can't measure total duration of multiple examples and use the result

----

From #653:

>> (I don't want to change the tests to GHC 8 because of how GHC 8 tests are structured).
>
> Can you elaborate?

If I use GHC 8 I'm forced to have my tests start with `Tests` and module name:

![image](https://user-images.githubusercontent.com/14909989/48319523-10492e80-e64a-11e8-9600-65f5fad0a9fb.png)

While it's always top level in GHC 7:

![image](https://user-images.githubusercontent.com/14909989/48319526-163f0f80-e64a-11e8-8393-aeda0a74dfff.png)

So I'm stuck with a particular form of test layout, which I think is a *loss of feature*.

> GHC 7 version had single entry point `main`. Why is this a problem now? `main = hspec spec`. You can have multiple groups of tests under it as before.

Because I can't do multiple top level blocks:

```haskell
main = do
  hspec $ do
    describe "Fixed tests" $ do
      it "encode" $ do
        ...
      it "decode" $ do
        ...
  hspec $ do
    describe "Random tests" $ do
      it "encode and decode" $ do
        ...
```

Having to always start `main` with `hspec spec` also disallows me to do advanced test setup usage (which some I've already been using already out there). Which I have to fix right now, and I haven't found out how to do them yet:

1. From [this kata](https://www.codewars.com/kata/one-line-task-how-many-digits): Perform actions with the source code before anything happens in the first place, so users can't use `unsafePerformIO` in their function to modify it during the first call to user function:

```haskell
main = do
  handle <- openFile "/home/codewarrior/solution.txt" ReadMode
  contents <- hGetContents handle
  putStr ("Your code: \n\n" ++ contents  ++ "\n\n")
  putStr ("Code length: " ++ (show $ length contents) ++ "\n")
  hClose handle
  
  hspec $ do
    describe "Tests" $ do
      it "Fixed tests" $ do
        ...
```

2. Custom timing measurement. I have a Haskell translation of [become immortal](https://www.codewars.com/kata/become-immortal) that I never got around to publish, that has tests structured like this:

```haskell
main = do
  hspec $ do
    describe "Fixed tests" $ do
      it "You should pass these before meeting the Elder" $ do
        ...
  start <- getCPUTime
  hspec $ do
    describe "The Elder is interested..." $ do
      it "\"Young man, you should learn a thing or two...\"" $ do
        putStrLn "<p><font color=\"green\">100 test cases\nm,n: 2^5 - 2^10\nl: 0 - 19\nt: 2^5 - 2^15</font></p>"
        quickCheckWith stdArgs { maxSuccess = 100 } $ do
          ...
  hspec $ do
    describe "The Elder is excited!" $ do
      it "\"You're too young and too simple!\"" $ do
        putStrLn "<p><font color=\"yellow\">300 test cases\nm,n: 2^8 - 2^20\nl: 0 - 9999\nt: 2^8 - 2^20</font></p>"
        quickCheckWith stdArgs { maxSuccess = 300 } $ do
          ...
  hspec $ do
    describe "The Elder is angry!" $ do
      it "\"And sometimes naive!\"" $ do
        putStrLn "<p><font color=\"red\">500 test cases\nm,n: 2^32 - 2^64\nl: 0 - 9999999\nt: 2^16 - 2^32</font></p>"
        quickCheckWith stdArgs { maxSuccess = 500 } $ do
          ...
  end <- getCPUTime
  let diff = (fromIntegral (end - start)) / (10^9) :: Double
  let str = if diff < 1000 then "You finished all the tests within 1 second. The Elder is very happy! +1s" else "The Elder is pleased!"
  hspec $ do
    describe ("You spent...: " ++ (show $ round diff) ++ " ms") $ do
      it str $ 1 `shouldBe` 1
```

There's a global `getCPUTime` that keep track of the total duration used to execute all the random tests, so then I can use it in the end to do some stuff according to its value. Values do not persist across `it` blocks, and you can't put `IO` actions until you've inside a `it` block.

Basically, by forcing `main = hspec spec` lots of flexibility is lost, and makes some katas untranslatable to Haskell even when there are nothing in the language that prevents it from being made. So it's kinda stupid.

**Comment by kazk** [comment]
> If I use GHC 8 I'm forced to have my tests start with `Tests` and module name:

I'll check if I can reduce the nesting.

---

> Because I can't do multiple top level blocks:

You should be able to.
```haskell
spec = do
  describe "Fixed tests" $ do
    it "encode" $ do
        ...
    it "decode" $ do
        ...
  describe "Random tests" $ do
    it "encode and decode" $ do
        ...
```
Doesn't this work?

---

> Having to always start within an `spec` also disallows me to do advanced test setup usage (which some I've already been using already out there). Which I have to fix right now, and I haven't found out how to do them yet:

Have you tried using [hooks](https://hspec.github.io/writing-specs.html#using-hooks)? I'll also look for other possible ways.

**Comment by Voileexperiments** [comment]
> You should be able to.
> ```
> spec = do
>  describe "Fixed tests" $ do
>    it "encode" $ do
>        ...
>    it "decode" $ do
>        ...
>  describe "Random tests" $ do
>    it "encode and decode" $ do
>        ...
> ```
> Doesn't this work?

No, there can only be one `describe` at top level block, unless it's inside *another* `describe` block:

```haskell
spec = do
  describe "Tests" $ do
    describe "Fixed tests" $ do
      it "encode" $ do
          ...
      it "decode" $ do
          ...
    describe "Random tests" $ do
      it "encode and decode" $ do
          ...
```

That's why I've resorted to using multiple `hspec`s.

**Comment by kazk** [comment]
It works, please try it.
GHC 7 one does have a bug that doesn't show the output correctly and that's probably why you used multiple `hspec`s.

**Comment by kazk** [comment]
> From this kata: Perform actions with the source code before anything happens in the first place, so users can't use unsafePerformIO in their function to modify it during the first call to user function:

For this one, you can use `beforeAll` with a function `getSolution :: IO String` or whatever you call it. This will read the file once before executing any tests and pass that to each spec.

```haskell
spec :: Spec
spec = beforeAll getSolution $ do
  it "Fixed tests" $ \_ -> do

  it "Code length test" $ \contents -> do
    (length contents) `shouldSatisfy` (<= 70)

  modifyMaxSuccess (const 50) $ it "Random tests" $ \_ -> do
```

**Comment by Voileexperiments** [comment]
> GHC 7 one does have a bug that doesn't show the output correctly and that's probably why you used multiple `hspec`s.

Oh yeah, now I remembered that was the reason I was using multiple `hspec`s.

> For this one, you can use `beforeAll` with a function `getSolution :: IO String` or whatever you call it. This will read the file once before executing any tests and pass that to each spec.

Thanks, I've fixed that kata and [another kata](https://www.codewars.com/kata/integer-to-nested-structure). The second one is kind of a big mess, not to mention that each `describe` block are only supposed to be run if the last test block did not have any failed tests (which I haven't put in the test code yet). I might get to it some time later.

**Comment by kazk** [comment]
I've removed the first "Tests" group so it's now just one level deeper than GHC 7 one and having the name of the module under test is reasonable.

For the last one, maybe something that measures timing can be added to `hspec-codewars`. It won't be exactly like the way you had it, but we might be able to do something similar.

I'm going to close this because other issues are resolved and we're not going back to using `hspec` function.

**Comment by kazk** [comment]
I got your [BECOME IMMORTAL](https://www.codewars.com/kata/become-immortal) working with GHC 8 by using `beforeAll` with `afterAll`, adjusting the type with `aroundWith` (stole from [solitude's kumite](https://www.codewars.com/kumite/5c703ea1c447bd4b75bafebf?sel=5c703ea1c447bd4b75bafebf)).
It shows different messages based on the duration of random tests.

**Comment by Voileexperiments** [comment]
@kazk There are two problems with the edit:

1. The subsequent `describe` blocks after the random tests is still run even if the ones before have failed. This is *not* desirable since the later tests are performance tests, which will cause timeout (and cause output to be not displayed because stdout is usually not flushed before container termination) if the user code is not performant enough. It's also intended in the other language versions.
2. The message is still shown even if some tests have failed.

So I guess I'll need to use `unsafePerformIO` and `IORef`s to keep track of the passed test amount manually and only execute the blocks and the final message if the numbers match?
